javapoet/src/main/java/com/palantir/javapoet/TypeSpec.java

```diff
@@ -66,6 +66,9 @@ public final class TypeSpec {
     private final Set<String> nestedTypesSimpleNames;
     private final List<Element> originatingElements;
     private final Set<String> alwaysQualifiedNames;
+    private final List<ParameterSpec> recordComponents;
+    private final boolean varargs;
+    private final MethodSpec compactConstructor;
```

Comment:
Motivated by https://github.com/square/javapoet/pull/840#issuecomment-975596085

Instead of having `recordComponents` and `varargs` here, I think it would be better to derive these from the `compactConstructor`. As suggested in that comment, it probably then makes sense to rename this to `recordConstructor`.

We can determine whether or not to emit the compact constructor by checking if the body is empty.